### PR TITLE
feat(aws-codebuild): add support for secondary sources and artifact in Projects

### LINKED ISSUE #738: CodeBuild: add support for secondarySources and secondaryArtifacts
I think this is quite a new feature in CodeBuild, but it would be great if CDK supported this (Currently I am blocked from using CDK to deploy our build projects for this reason)

https://docs.aws.amazon.com/codebuild/latest/userguide/sample-multi-in-out.html

### PR DESCRIPTION
**Pull request by skinny85**
This also adds support for multiple input and output artifacts in the CodeBuild
CodePipeline Actions.

BREAKING CHANGE: this changes the way CodeBuild Sources are constructed
(we moved away from multiple parameters in the constructor,
in favor of the more idiomatic property interface).

Fixes #738.

-------------------------------------------------------------

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 1]
```diff
@@ -119,3 +124,103 @@ To define CloudWatch event rules for build projects, use one of the `onXxx` meth
...
+and produce multiple outputs. For example:
+
+```ts
+const project = new codebuild.Project(this, 'MyProject', {
+    secondarySources: [
+        new codebuild.CodeCommitSource({
+            identifier: 'source2',
+            repository: repo,
+        }),
+    ]
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/README.md]: Missed a comma here, will fix.

**Comment by skinny85** [comment]
@eladb ping ping when you have some time :)

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 2]
```diff
@@ -14,7 +14,9 @@ import codecommit = require('@aws-cdk/aws-codecommit');
 
 const repo = new codecommit.Repository(this, 'MyRepo', { repositoryName: 'foo' });
 new codebuild.Project(this, 'MyFirstCodeCommitProject', {
-    source: new codebuild.CodeCommitSource(repo)
+    source: new codebuild.CodeCommitSource({
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/README.md]: The indentation here is off. Our standard these days is 2 spaces.

This is probably true for all files?

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 3]
```diff
@@ -119,3 +124,103 @@ To define CloudWatch event rules for build projects, use one of the `onXxx` meth
...
+        phases: {
+            build: {
+                commands: [
+                    'cd $CODEBUILD_SRC_DIR_source2',
+                    'touch output2.txt',
+                ],
+            }
+        },
+        artifacts: {
+            'secondary-artifacts': {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/README.md]: OMFG the backwards compatible API extensions they pain me so much 

oh no

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 4]
```diff
@@ -119,3 +124,103 @@ To define CloudWatch event rules for build projects, use one of the `onXxx` meth
...
+                },
+            },
+        },
+    },
+});
+```
+
+#### Multiple inputs and outputs in CodePipeline
+
+Unfortunately, you cannot use neither `secondarySources` nor `secondaryArtifacts`
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/README.md]: None of y'all can't use none of your nothings!

But seriously, is the double negation here correct? "Can not + neither"? Feels to me like it should be "cannot use either"... ?

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 5]
```diff
@@ -119,3 +124,103 @@ To define CloudWatch event rules for build projects, use one of the `onXxx` meth
...
+const sourceAction1 = repository1.addToPipeline(sourceStage, 'Source1');
+const sourceAction2 = repository2.addToPipeline(sourceStage, 'Source2', {
+    outputArtifactName: 'source2',
+});
+
+const buildStage = pipeline.addStage('Build');
+const buildAction = project.addBuildToPipeline(buildStage, 'Build', {
+    inputArtifact: sourceAction1.outputArtifact,
+    outputArtifactName: 'artifact1', // for better buildspec readability - see below
+    additionalInputArtifacts: [
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/README.md]: Wait, I'm confused. First you're saying "unfortunately this doesn't work" but now you're showing me code that does exactly what I want?

Is the "unfortunate" part of it that they're named differently?

File: packages/@aws-cdk/aws-codebuild/README.md [hunk 6]
```diff
@@ -119,3 +124,103 @@ To define CloudWatch event rules for build projects, use one of the `onXxx` meth
...
+        sourceAction2.outputArtifact, // this is where 'source2' comes from
+    ],
+    additionalOutputArtifactNames: [
+        'artifact2',
+    ],
+});
+// use buildAction.additionalOutputArtifacts as inputs to later stages...
+```
+
+**Note**: when a CodeBuild Action in a Pipeline has more than one input,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/README.md]: Is this section trying to say that you need to list both `artifact1` and `artifact2` under `secondary-artifacts`? An example might help to solidify that. I'm trying hard to parse this text but I'm finding it hard.

File: packages/@aws-cdk/aws-codebuild/lib/pipeline-actions.ts [hunk 7]
```diff
@@ -45,13 +63,23 @@ export class PipelineBuildAction extends codepipeline.BuildAction {
...
     setCodeBuildNeededPermissions(props.stage, props.project, true);
+
+    handleAdditionalInputOutputArtifacts(props, this,
+      // pass functions to get around protected members
+      (artifact) => this.addInputArtifact(artifact),
+      (artifactName) => this.addOutputArtifact(artifactName));
+  }
+
+  public additionalOutputArtifacts(): codepipeline.Artifact[] {
+    return this._outputArtifacts.slice(1);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/lib/pipeline-actions.ts]: Should we be able to address these by name?

I.e., probably add a function?

```ts
additionalOutputArtifact(name: string): codepipeline.Artifact | Exception
```

Feels like if I want to take one of 5 output artifacts to pass into another step, there should be a convenient way for me to get that object.

(Yes, I know it's just a string wrapper and it feels silly to use a string to get a typed string to pass onto another object that in principle only expected that string THAT YOU ALREADY HAD. So it goes :))

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 8]
```diff
@@ -532,6 +566,22 @@ export class Project extends ProjectRef {
...
   }
 
+  /**
+   * @override
+   */
+  public validate(): string[] {
+    const ret = new Array<string>();
+    if (this.source.type === SourceType.CodePipeline) {
+      if (this._secondarySources.length > 0) {
+        ret.push('A Project with a CodePipeline Source cannot have secondary sources');
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: Nice check!

Can we give the user something actionable to do here to correct their (most likely) problem?

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 9]
```diff
@@ -542,6 +592,22 @@ export class Project extends ProjectRef {
     }
   }
 
+  public addSecondarySource(secondarySource: BuildSource): void {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: Public method needs a docstring eh

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 10]
```diff
@@ -542,6 +592,22 @@ export class Project extends ProjectRef {
...
+    if (!secondarySource.identifier) {
+      throw new Error('The identifier attribute is mandatory for secondary sources');
+    }
+    secondarySource._bind(this);
+    this._secondarySources.push(secondarySource);
+  }
+
+  public addSecondaryArtifact(secondaryArtifact: BuildArtifacts): any {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: Public

File: packages/@aws-cdk/aws-codebuild/lib/artifacts.ts [hunk 11]
```diff
@@ -2,26 +2,80 @@ import s3 = require('@aws-cdk/aws-s3');
...
+  }
+
+  public _bind(_project: Project) {
     return;
   }
+
+  public toArtifactsJSON(): cloudformation.ProjectResource.ArtifactsProperty {
+    const artifactsProp = this.toArtifactsProperty();
+    return {
+      artifactIdentifier: this.identifier,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/lib/artifacts.ts]: If people supply an `artifactIdentifier` for the primary source, does it blow up?

Might want to be invariant to that.

File: packages/@aws-cdk/aws-codebuild/lib/source.ts [hunk 12]
```diff
@@ -6,73 +6,150 @@ import { cloudformation } from './codebuild.generated';
...
+  }
+
+  public _bind(project: Project) {
+    this.bucket.grantRead(project.role);
+  }
+
+  protected toSourceProperty(): any {
     return {
-      type: SourceType.CodePipeline
+      location: new cdk.FnConcat(this.bucket.bucketName, '/', this.path)
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codebuild/lib/source.ts]: We can do TypeScript string interpolation these days and the world is great:

```
location: `${this.bucket.bucketName}/${this.path}`
```

File: packages/@aws-cdk/aws-codebuild/README.md (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/README.md]: Yes, none of the ReadMe files were changed to 2 spaces when the indentation was changed for code, hence I'm sticking to 4 here. We should probably do a separate pass over the Readmes to fix them all.

File: packages/@aws-cdk/aws-codebuild/README.md (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/README.md]: Lol :)

File: packages/@aws-cdk/aws-codebuild/README.md (hunk 4 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/README.md]: You are correct sir. TIL. Changed to "either or".

File: packages/@aws-cdk/aws-codebuild/README.md (hunk 5 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/README.md]: I meant to convey that you can't use the `secondarySources` / `secondaryArtifacts` properties in a Pipeline Project, but you can use `additionalInputArtifacts` / `additionalOutputArtifacts`. Obviously I failed. I'll reword the explanation.

File: packages/@aws-cdk/aws-codebuild/README.md (hunk 6 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/README.md]: Yes. Explanation fail as well. Will add an example and re-word.

File: packages/@aws-cdk/aws-codebuild/lib/project.ts (hunk 8 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: Sure - probably mentioning the `additionalInputArtifacts` / `additionalOutputArtifacts` properties of the CodePipeline Actions would be good enough here?

File: packages/@aws-cdk/aws-codebuild/lib/project.ts (hunk 9 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: Yep. Will add.

File: packages/@aws-cdk/aws-codebuild/lib/project.ts (hunk 10 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: ... method needs a docstring. Yep.

File: packages/@aws-cdk/aws-codebuild/lib/artifacts.ts (hunk 11 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/artifacts.ts]: Nope, that's fine - it won't cause any problems.

File: packages/@aws-cdk/aws-codebuild/README.md (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/README.md]: Just wait until CDK hits 1.0, and we'll have to do crap like this ourselves 😀

File: packages/@aws-cdk/aws-codebuild/lib/source.ts (hunk 12 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/source.ts]: Amazing 😱

File: packages/@aws-cdk/aws-codebuild/lib/pipeline-actions.ts (hunk 7 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/pipeline-actions.ts]: If I thought that was silly, I wouldn't be in this business 😀. Added the method!

**Comment by skinny85** [comment]
Incorporated @rix0rrr 's comments.
